tests/virt_autotest/login_console_as_kvm_or_xen_role.pm

```diff
@@ -0,0 +1,55 @@
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.  This file is offered as-is,
+# without any warranty.
+#
+# Summary: For openSUSE virtualization test only. login in console, install kvm/xen patterns if needed.
+# Maintainer: Julie CAO <jcao@suse.com>
+
+use strict;
+use warnings;
+use base 'virt_autotest_base';
+use testapi;
+use ipmi_backend_utils;
+use Utils::Backends qw(use_ssh_serial_console is_remote_backend);
+use utils qw(zypper_call systemctl);
+use virt_autotest::utils qw(is_kvm_host is_xen_host);
+
+sub run {
+
+    #skip TW host installation and directly login
+    if (get_var('IPMI_DO_NOT_RESTART_HOST')) {
+        select_console 'sol', await_console => 0;
+        send_key 'ret';
+    }
+
+    use_ssh_serial_console unless check_screen('text-logged-in-root');
+
+    return if get_var('IPMI_DO_NOT_RESTART_HOST');
+
+    die 'Need one of both to be true: is_kvm_host || is_kvm_host' unless is_kvm_host || is_xen_host;
+    my $hypervisor = is_kvm_host ? 'kvm' : 'xen';
+    zypper_call('--gpg-auto-import-keys ref');
+    zypper_call("in -t pattern ${hypervisor}_server ${hypervisor}_tools", 1800);
+    set_serial_console_on_vh('', '', $hypervisor);
+
+    virt_autotest::utils::install_default_packages();
+
+    #libvirtd failed to be started in Tumbleweed
+    #not sure if it is a bug yet, just work it around here to make tests run
+    systemctl 'status libvirtd', ignore_failure => 1;
+    save_screenshot;
+    script_run "echo DEBUG:; echo PATH=\$PATH; which dnsmasq; which iptables; which ip6tables; which ebtables";
+    save_screenshot;
+    zypper_call('in dnsmasq iptables bridge-utils', 1800);
+    assert_script_run("ln -s /usr/sbin/$_ /sbin/$_") for qw(dnsmasq iptables ip6tables ebtables);
+    systemctl 'restart libvirtd', ignore_failure => 1;
+    systemctl 'status libvirtd', ignore_failure => 1;
+    systemctl 'enable libvirtd', ignore_failure => 1;
+    save_screenshot;
+}
+
+sub test_flags { {fatal => 1} }
+
+1;
```

Comment:
Why this new login_console module for opensuse is quite necessary ? To be honest, I can not see very new things here compared with already existing modules, for example, login_console+uefi_guest_installation module. Or do you want to get rid of old login_consle/install_packages/update_packages totally ?

Reply:
I am going to rename this module to setup_hypervisor.pm(or other name) as this name is misleading. 
Yes, install_package or update_package have not been needed any more in TW tests, because UEFI guests installation and subsequent tests need not either QA packages or any updates, or even build number replacement and enabling debug log.

Reply:
And in which order that this setup_hypervisor will be called ? As I said in last comment, it seems that there is very few new things in this setup_hypervisor.pm anyway, so why do you need this module ? IMHO, all things in this module can be provided by the  already exiting login_console and uefi_guest_installation. Hope you can help clarify your intention to do so.

Reply:
login_console is not needed in TW as there is no login as kvm/xen role. 

The new setup_hypervisor.pm is to switch to ssh console, zypper install kvm/xen patterns, set serial console for kvm/xen, make some necessary workaround.  Have all these things been implemented in uefi_guest_installation? I did not notice that.

Reply:
Well, `login_console` is needed in TW as you log in there to the console too.

I'm OK with a new module for installing the KVM / XEN patterns including the necessary console configuration.

Reply:
> login_console is not needed in TW as there is no login as kvm/xen role.
> 
> The new setup_hypervisor.pm is to switch to ssh console, zypper install kvm/xen patterns, set serial console for kvm/xen, make some necessary workaround. Have all these things been implemented in uefi_guest_installation? I did not notice that. And these part should not be done in uefi_guest_installation as host reboot is needed in TW because kvm/xen login is required.
> 
> The order it is called please check in the validation run, or http://10.67.129.51/tests/2501
> install host => reboot_after_installation => reconnect_mgmt_console => first_boot => setup_hypervisor.pm => reboot_and_wait_up_normal => uefi_guest_installation => ....

I just think uefi_guest_installation and its base module already  installs kvm or xen server role, so I think you do not need a new module to serve this purpose, so login_console and uefi_guest_installation enough. I did not suggest implement anything out of box in uefi_guest_installation. I just want to know why you need a brand new module, so hope your detailed explanation, for example, "because no module implements this anywhere but it is really required in a new module". Could you help give more comment on your new module's introduction and its role and purpose in the test flow and why it is really needed ?

Reply:
put comments at the beginning of the module(switch_to_ssh_and_install_hypervisor.pm):
# Summary: For openSUSE virtualization test only. login in console, install kvm/xen patterns if needed.
#  - Even if you'd like to run tests without host installation(IPMI_DO_NOT_RESTART_HOST=1), this module is still necessary as login console in this module is required.
#  - This module is added for openSUSE TW because of the difference beteen SLE and TW. Meanwile, login_console, install_package and update_package from SLE are not needed. The reasons are listed below:
#    -- login_console is not needed in TW as there is no login as kvm/xen role. reconnect_mgmt_console and first_boot take care of the login function. 
#    -- have to zypper install kvm/xen patterns in TW
#    -- no QA packages have been required any more.
#    -- no update is needed as it is Tumbleweed. 
#    -- reboot is needed after kvm/xen patterns are installed so the module must run before reboot_and_wait_up_normal.

Reply:
But I think "kvm or xen boot entry" should become available after installing kvm_server or xen_server  right ? Or there will no such boot entry anyway for good.

Reply:
Yes, xen boot entry becomes available after zypper install xen patters. So the reboot_and_wait_up_normal.pm from SLE which calls login_console works in TW.